pennylane/ops/qubit/state_preparation.py

```diff
@@ -69,15 +68,51 @@ class BasisState(StatePrepBase):
     [0.+0.j 0.+0.j 0.+0.j 1.+0.j]
     """
 
-    num_wires = AnyWires
-    num_params = 1
-    """int: Number of trainable parameters that the operator depends on."""
+    def __init__(self, features, wires, id=None):
 
-    ndim_params = (1,)
-    """int: Number of dimensions per trainable parameter of the operator."""
+        if isinstance(features, list):
+            features = qml.math.stack(features)
+
+        tracing = qml.math.is_abstract(features)
+
+        if qml.math.shape(features) == ():
+            if not tracing and features >= 2 ** len(wires):
+                raise ValueError(
+                    f"Features must be of length {len(wires)}, got features={features} which is >= {2 ** len(wires)}"
+                )
+            bin = 2 ** math.arange(len(wires))[::-1]
+            features = qml.math.where((features & bin) > 0, 1, 0)
+
+        wires = Wires(wires)
+        shape = qml.math.shape(features)
+
+        if len(shape) != 1:
+            raise ValueError(f"Features must be one-dimensional; got shape {shape}.")
+
+        n_features = shape[0]
+        if n_features != len(wires):
+            raise ValueError(
+                f"Features must be of length {len(wires)}; got length {n_features} (features={features})."
+            )
+
+        if not tracing:
+            features_list = list(qml.math.toarray(features))
```

Comment:
Maybe we could use something like `qml.math.allclose(qml.math.unique(features), [0, 1])`